### PR TITLE
Add the feature for automatically adding attachments with tracking function

### PR DESCRIPTION
**Pull request by jli53**
Hi Gophish Community,

I added a feature that allows Gophish to automatically attach html or doc file with tracking function. It is, when users open the html file or open doc file with macro enabled, Gophish server will get notification and display results. In the project directory there is a "b.doc" file which has vba script inside. The only thing need to modify is the url in vba script, which should be the ip address of Gophish server. When correct the ip address, admin can check the checkbox for tracking attachments and choose whatever name for the attachments, then server will generate attachments and attach them with emails sent out.

If you have any questions for my code, please feel free to give me a message.

Thanks,
Jiaming

### CODE REVIEW

**Comment by jordan-wright** [comment]
Hey @jli53 - thanks for this!

Just a heads up, I will be working to review the changes as soon as possible. Comments will be heading your way soon.

Thanks again!

File: auth/auth.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ import (
 
 	"crypto/rand"
 
-	"github.com/gophish/gophish/models"
+	"github.com/jli53/gophish/models"
```
Comment (jordan-wright) [inline auth/auth.go]: All these dependencies will need to be reverted to gophish/gophish - I know it's a pain to work with when forking.

File: config.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
 	"admin_server" : {
-		"listen_url" : "127.0.0.1:3333",
+		"listen_url" : "0.0.0.0:3333",
```
Comment (jordan-wright) [inline config.json]: This should not be changed in this pull request - the default admin will need to stay at 127.0.0.1

File: models/models.go [hunk 3]
```diff
@@ -34,6 +34,8 @@ const (
 	EVENT_OPENED         string = "Email Opened"
 	EVENT_CLICKED        string = "Clicked Link"
 	EVENT_DATA_SUBMIT    string = "Submitted Data"
+	EVENT_HTML_OPENED	 string = "HTML Opened"
```
Comment (jordan-wright) [inline models/models.go]: You'll need to run ```gofmt``` on the code to ensure proper formatting.

File: static/js/app/campaign_results.js [hunk 4]
```diff
@@ -2,6 +2,20 @@ var map = null
...
 // statuses is a helper map to point result statuses to ui classes
 var statuses = {
+    "HTML Opened": {
+        slice: "ct-slice-donut-opened",
+        legend: "ct-legend-opened",
+        label: "label-danger",
+        icon: "fa-file",
+        point: "ct-point-opened"
+    },
+    "Doc Opened": {
```
Comment (jordan-wright) [inline static/js/app/campaign_results.js]: "Doc" is pretty informal - we'll want to change this to "attachment" or something more official sounding.

File: util/util.go [hunk 5]
```diff
@@ -96,3 +96,11 @@ func ParseCSV(r *http.Request) ([]models.Target, error) {
 	}
 	return ts, nil
 }
+
+func GetDoc(rid string) []byte{
```
Comment (jordan-wright) [inline util/util.go]: Again - you'll need to ```go fmt```.

**Comment by jordan-wright** [comment]
A couple of other things:

Please don't add the gophish binary to version control - we publish official binaries that we compile from the source.

I'm not sure I want to go the route of managing ```.doc``` files, since they are an odd binary format. I would prefer to handle proper parsing and extraction of ```.docx``` files.

File: static/js/app/campaign_results.js (hunk 4 above)
Comment (jli53) [inline static/js/app/campaign_results.js]: There are two kinds of attachments in the program, which are html and ms word. One of them is called "HTML Opened". Do you think it is OK if I changed "Doc Opened" to "WORD Opened"?

**Comment by jli53** [comment]
Hi Jordan,

Thanks for reviewing my code. I correct most problems you pointed out.

I also hated to deal with ".doc". However, I checked some more details about ".docx" and found that ".docx" does not support script, instead it is ".docm" that supports script. After extracting ".docm" file, there is a odd binary file which corresponds to the script part. So if we are going to do MS Word attachment tracking, we have to deal with ".doc" or ".docm".

File: gophish.go [hunk 6]
```diff
@@ -32,10 +32,10 @@ import (
 	"os"
 	"sync"
 
-	"github.com/gophish/gophish/config"
-	"github.com/gophish/gophish/controllers"
-	"github.com/gophish/gophish/models"
 	"github.com/gorilla/handlers"
+	"github.com/jli53/gophish/config"
```
Comment (jordan-wright) [inline gophish.go]: These dependencies are still pointing to the wrong place - should be ```github.com/gophish/gophish```.

**Comment by leonteale** [comment]
has this been implemented yet. i would love for this to be a feature ! :D

**Comment by hpscrappy** [comment]
is there any progress on this yet, i am interested in leveraging this.

**Comment by hpscrappy** [comment]
the files will not attach to the emails in the campaign. I believe it is an issue with the master branch also.

**Comment by jli53** [comment]
Hey, which type of files you tried? HTML or doc? I'll check it soon. Thanks.

**Comment by hpscrappy** [comment]
i tried both separately, it will not even attach an uploaded .txt attachment, but send all on the test e-mail.

**Comment by ustayready** [comment]
This is actually perfect timing, I was just getting ready to add a tracking mechanism for macros. Currently, I use macros that retrieve and pass machine specific data elements to a remote URL. I grab that machine name and userid, assuming its available, so I can trace the macro to machines within the network.

**Comment by ustayready** [comment]
I think I have an idea for a solution. If we can set the tracking url in a cell on a sheet using something like https://github.com/tealeg/xlsx then have the macro read from the cell when generating the URL, then attach the xlsx to the email before sending. Thoughts?

**Comment by ustayready** [comment]
I don't mind giving it a shot but I am new to golang, might be a little slower.

**Comment by jli53** [comment]
My idea was adding an id at the end of doc file and write a macro
reading the id and sending a request to server with the id. Then server can
tell which user open the doc attachment.

**Comment by ustayready** [comment]
Ahh ok, how would you add the id to the end of the doc? Like the actual document or the filename?

**Comment by hpscrappy** [comment]
any progress on this yet, or is it just me.

**Comment by leonteale** [comment]
Assuming they have or enable macros?

**Comment by ustayready** [comment]
Looks like after I pulled jli53:master it broke my attachments. I blew away the master and rebuilt using go get yet attachments are still not working. Might be an environment thing, just figured I'd drop a note before others merge until it's resolved.

**Comment by jli53** [comment]
Oh. My branch works before. However recently I tried to merge remote master
branch and some issues happened. I haven't figured them out yet but I'll
keep working on it recently.

**Comment by hpscrappy** [comment]
sounds good thanks

**Comment by hpscrappy** [comment]
has this been fixed since the master update

**Comment by jli53** [comment]
Not yet. I got problem in launching campaign function. When I created a
campaign and click launch, the button just doesn't work. Still need further
check to see what's wrong with my code.

Sorry for the delay, but I only have time on weekends. Hopefully, I can
figure it out this week. Thanks.

**Comment by jli53** [comment]
Hey I just finished testing the code on my machine and it works well, including the tracking attachment function. Hey Jordan, can you please check my code again and tell me if there are something else I need to modify. Thanks.

**Comment by hpscrappy** [comment]
seems to be working better, the only issue i have is with the b.doc 
image

**Comment by jli53** [comment]
I forgot to mention three very important points about b.doc:

1. The file name has to be exact b.doc because in vbscript the file name is
hard coded as "b.doc". (Did you change the file name so you saw that error?
If not, please send the b.doc to me via email and I'll check it, thanks)

2. You need to change the IP address in VBscript to your gophish server IP
before you send it. The best way to change is via ms word. In your
situation, if you click debug, you'll see the script. Modifying b.doc via
binary tools or txt tools is very easy to destroy the script inside.

3.this function only works for Windows now because the script only works
there.

**Comment by hpscrappy** [comment]
i have only changed the server information in the script, in the example above everything else is the same and i did use the auto genereated name

**Comment by jli53** [comment]
I test it on my machine again and it works. Can you please send me the "b.doc" via email?(shadow4walker@gmail.com) I'll check it ASAP. Thanks.

**Comment by blackwat3r88** [comment]
Hi jli53, when I start the campaign, every email has two attachments, b.doc and b.doc.auto ( size 0 byte ). Opening the file b.doc, macro generates an error because is reading out file and checking the diff between hex of the original b.doc and the b.doc attached are the same, and probably the attachments haven't the RId into.

**Comment by jordan-wright** [comment]
Hi @jli53 - I'd like to apologize for the radio silence on this one.

This is looking good. I was concerned at first that this isn't _exactly_ what I was thinking for attachment support in terms of how the workflow looks, etc.

*But* - good news. I just realized I can change the base branch so now you aren't merging into master, which lets me build from your changes while still letting you keep the credit. I'll go ahead and merge this now so I can build on what you started.

**Comment by jli53** [comment]
Hey @jordan-wright , thanks for replying. That sounds great! If you have any problems on my implementation, please let me know! And @blackwat3r88 , I never saw that issue before. Let me test it on another machine this weekend. Thanks.

**Comment by santhoshmla** [comment]
Hi @jli53 Thanks for the attachment tracking functionality
testing html attachments, assuming i used an html attachment with file name test123.html,
Every sent email has two attachments eg. test123.html & test123.html.auto

attachment test123.html is the original file that i attached, it does not have tracking rid in it.

have you tested html attachments? is it working as expected on your end?

**Comment by jli53** [comment]
Another user also told me this issue. Give me one more weekend and I'll let you know when I figure it out. Thanks.

**Comment by jli53** [comment]
Hi guys, @hpscrappy @blackwat3r88 @santhoshmla 

I think I figured out what happened :)

I think you guys uploaded files callled "a.html" or "b.doc" manually in Template page, also with two checkbox checked. That's why you have "a.html" and "a.html.auto" in every email because one is attached by you, the other one is automatically attached by Gophish itself due to the checked checkbox. And the related id will only be attached to the attachments attached by Gophish. If what i'm guessing is correct, to make it works on your machine, all you need to do is check checkbox on Template page, NO NEED TO upload any files. Then you will only see "a.html.auto" and "b.html.auto". And they will work with tracking functionality. You can change the filename for html in textfield, but not doc file for now.

Sorry for the confusion. Hope you are enjoying long weekends.

**Comment by kyra007** [comment]
Hi Jli53, Jordan and the rest of community.

Great enhancement, you guys are making great contributions to cyber-threat prevention community, congrats! 
Jli53, I do need some further elaboration though. I may be missing the point and purpose of auto generated files b.doc.auto or a.html.auto. For example, when building a phishing campaign, having a word macro file attached, one (target/educating victim) would only expect the final (rid equipped; for tracking) macro file in delivered e-mail message.

I tested different combinations (having the "Add doc Attachment" option checked or without), and I always receive the b.doc.auto (only) attached to delivered mail. My understanding was, the targeted user is supposed to receive the b.doc "macro enabled" file, that has its unique RID URL to be called automatically (from embedded VBA script, hitting gophish server), after having user clicked on "Allow macros ..."? 

Thank you for your clarification!!

Regards,
kyra007

**Comment by ShivaPoola143** [comment]
Hi Jil, How should i do. I am a beginners. What i put/write in .doc file.

Will you please explain briefly.

**Pull request merged**